test/OpenTelemetry.Tests/Impl/Metrics/Aggregators/CounterAggregatorTest.cs

```diff
@@ -0,0 +1,130 @@
+// Copyright 2018, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+using System;
+using System.Threading;
+using OpenTelemetry.Metrics.Aggregators;
+using Xunit;
+
+namespace OpenTelemetry.Metrics.Test
+{
+    public class CounterAggregatorTest
+    {
+        [Fact]
+        public void CounterAggregatorAggregatesCorrectlyWhenMultipleThreadsUpdatesLong()
+        {
+            // create an aggregator
+            CounterSumAggregator<long> aggregator = new CounterSumAggregator<long>();
+            var sum = aggregator.ValueFromLastCheckpoint();
+
+            // we start with 0.
+            Assert.Equal(0, sum);
+
+            // setup args to threads.
+            var mre = new ManualResetEvent(false);
+            var args = new Tuple<ManualResetEvent, CounterSumAggregator<long>>(mre, aggregator);
+            Thread[] t = new Thread[10];
+            for (int i = 0; i < 10; i++)
+            {
+                t[i] = new Thread(LongMetricUpdateThread);
+                t[i].Start(args);
+            }
+
+            // kick-off all the threads.
+            mre.Set();
+
+            for (int i = 0; i < 10; i++)
+            {
+                // wait for all threads to complete
+                t[i].Join();
+            }
+
+            // check point.
+            aggregator.Checkpoint();
+            sum = aggregator.ValueFromLastCheckpoint();
+
+            // 1000000 times 10 by each thread. times 10 as there are 10 threads
+            Assert.Equal(100000000, sum);
+        }
+
+        [Fact]
+        public void CounterAggregatorAggregatesCorrectlyWhenMultipleThreadsUpdatesDouble()
+        {
+            // create an aggregator
+            CounterSumAggregator<double> aggregator = new CounterSumAggregator<double>();
+            var sum = aggregator.ValueFromLastCheckpoint();
+
+            // we start with 0.0
+            Assert.Equal(0.0, sum);
+
+            // setup args to threads.
+            var mre = new ManualResetEvent(false);
+            var args = new Tuple<ManualResetEvent, CounterSumAggregator<double>>(mre, aggregator);
+            Thread[] t = new Thread[10];
+            for (int i = 0; i < 10; i++)
+            {
+                t[i] = new Thread(DoubleMetricUpdateThread);
+                t[i].Start(args);
+            }
+
+            // kick-off all the threads.
+            mre.Set();
```

Comment:
It's likely this will be hit before any thread is even started. 
Perhaps 2 `ManualResetEvent`s are needed here. One to block this method and one to block inside `DoubleMetricUpdateThread` until the last thread (using a counter) spawns.
Perhaps a counter (Interlock.Increment?) and when theadCount is hit, `mre.Set` in the `DoubleMetricUpdateThread` to unblock the test, whick hits `Set` to unblock all threads.

If that makes any sense :)

Reply:
Absolutely!

Reply:
done.